app/code/core/Mage/Log/Model/Visitor.php

```diff
@@ -101,17 +101,22 @@ public function __construct(array $data = [])
     protected function _construct()
     {
         $this->_init('log/visitor');
-        $userAgent = $this->_httpHelper->getHttpUserAgent();
+        if ($this->_logCondition->isLogDisabled()) {
+            $this->_skipRequestLogging = true;
+            return;
+        }
+
         $ignoreAgents = $this->_config->getNode('global/ignore_user_agents');
         if ($ignoreAgents) {
             $ignoreAgents = $ignoreAgents->asArray();
-            if (in_array($userAgent, $ignoreAgents)) {
-                $this->_skipRequestLogging = true;
+            $userAgent = $this->_httpHelper->getHttpUserAgent();
+            foreach ($ignoreAgents as $ignoreAgent) {
+                if (stripos($userAgent, $ignoreAgent) !== false) {
+                    $this->_skipRequestLogging = true;
+                    break;
+                }
```

Comment:
The longer the list, the slower it gets. I am weary.

Reply:
Can a regexp help to improve performance?

Reply:
regex is probably worst. I don't have any idea how to do this faster.

Reply:
> regex is probably worst. I don't have any idea how to do this faster.

Also expected regex to be slower, but it doesn't seem so.

```php
$ignoreAgents = range(0, 500000);
$ignoreAgents = array_map('strval', $ignoreAgents);
$ignoreAgents[] = 'google';
$userAgent = 'Googl'; #test: google, Google, Googl

function isIgnoredUnserAgent(string $userAgent, array $ignoreAgents): bool
{
    foreach ($ignoreAgents as $ignoreAgent) {
        if (stripos($userAgent, $ignoreAgent) !== false) {
            return true;
        }
    }
    return false;
}

function isIgnoredUnserAgentRegex(string $userAgent, array $ignoreAgents): bool
{
    $ignoreAgent = implode(' ', $ignoreAgents);
    return preg_match("/\b{$userAgent}\b/i", $ignoreAgent) === 1;
}

var_dump(isIgnoredUnserAgent($userAgent, $ignoreAgents));
var_dump(isIgnoredUnserAgentRegex($userAgent, $ignoreAgents));
```

## isIgnoredUnserAgent
<table rules="rows" width="90%" cellspacing="1" cellpadding="2" bordercolor="#bdc7d8" border="1" align="center">
<tbody><tr bgcolor="#bdc7d8" align="right"><th align="left"><nobr><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=isIgnoredUnserAgent&amp;source=ddev&amp;sort=fn">Function Name</a></nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=isIgnoredUnserAgent&amp;source=ddev&amp;sort=ct">Calls</a></nobr></th><th class="vwbar"><nobr>Calls%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=isIgnoredUnserAgent&amp;source=ddev&amp;sort=wt">Incl. Wall Time<br>(microsec)</a></nobr></th><th class="vwbar"><nobr>IWall%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=isIgnoredUnserAgent&amp;source=ddev&amp;sort=cpu">Incl. CPU<br>(microsecs)</a></nobr></th><th class="vwbar"><nobr>ICpu%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=isIgnoredUnserAgent&amp;source=ddev&amp;sort=mu">Incl.<br>MemUse<br>(bytes)</a></nobr></th><th class="vwbar"><nobr>IMemUse%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=isIgnoredUnserAgent&amp;source=ddev&amp;sort=pmu">Incl.<br> PeakMemUse<br>(bytes)</a></nobr></th><th class="vwbar"><nobr>IPeakMemUse%</nobr></th></tr><tr bgcolor="#e0e0ff"><td><b><i><center>Current Function</center></i></b></td></tr><tr><td><a href="">isIgnoredUnserAgent</a></td><td class="vbar">1</td>
<td class="vbar">0.0%</td>
<td class="vbbar">452,966</td>
<td class="vbbar">28.0%</td>
<td class="vbar">450,998</td>
<td class="vbar">28.1%</td>
<td class="vbar">1,120</td>
<td class="vbar">0.0%</td>
<td class="vbar">0</td>
<td class="vbar">0.0%</td>
</tr><tr bgcolor="#ffffff"><td style="text-align:right;color:blue">Exclusive Metrics  for Current Function</td><td class="vbar"></td><td class="vbar"></td><td type="Child" metric="wt" class="vbbar">281,408</td>
<td type="Child" metric="wt" class="vbbar">62.1%</td>
<td type="Child" metric="cpu" class="vbar">214,517</td>
<td type="Child" metric="cpu" class="vbar">47.6%</td>
<td type="Child" metric="mu" class="vbar">592</td>
<td type="Child" metric="mu" class="vbar">52.9%</td>
<td type="Child" metric="pmu" class="vbar">0</td>
<td type="Child" metric="pmu" class="vbar">N/A%</td>
</tr><tr bgcolor="#e0e0ff"><td><b><i><center>Parent function</center></i></b></td></tr><tr><td><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=main%28%29&amp;source=ddev">main()</a></td><td type="Parent" metric="ct" class="vbar">1</td>
<td type="Parent" metric="ct" class="vbar">100.0%</td>
<td type="Parent" metric="wt" class="vbbar">452,966</td>
<td type="Parent" metric="wt" class="vbbar">100.0%</td>
<td type="Parent" metric="cpu" class="vbar">450,998</td>
<td type="Parent" metric="cpu" class="vbar">100.0%</td>
<td type="Parent" metric="mu" class="vbar">1,120</td>
<td type="Parent" metric="mu" class="vbar">100.0%</td>
<td type="Parent" metric="pmu" class="vbar">0</td>
<td type="Parent" metric="pmu" class="vbar">N/A%</td>
</tr><tr bgcolor="#e0e0ff"><td><b><i><center>Child function</center></i></b></td></tr><tr><td><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=stripos&amp;source=ddev">stripos</a></td><td type="Child" metric="ct" class="vbar">500,002</td>
<td type="Child" metric="ct" class="vbar">100.0%</td>
<td type="Child" metric="wt" class="vbbar">171,558</td>
<td type="Child" metric="wt" class="vbbar">37.9%</td>
<td type="Child" metric="cpu" class="vbar">236,481</td>
<td type="Child" metric="cpu" class="vbar">52.4%</td>
<td type="Child" metric="mu" class="vbar">528</td>
<td type="Child" metric="mu" class="vbar">47.1%</td>
<td type="Child" metric="pmu" class="vbar">0</td>
<td type="Child" metric="pmu" class="vbar">N/A%</td>
</tr></tbody></table>

## isIgnoredUnserAgentRegex
<table rules="rows" width="90%" cellspacing="1" cellpadding="2" bordercolor="#bdc7d8" border="1" align="center">
<tbody><tr bgcolor="#bdc7d8" align="right"><th align="left"><nobr><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=isIgnoredUnserAgentRegex&amp;source=ddev&amp;sort=fn">Function Name</a></nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=isIgnoredUnserAgentRegex&amp;source=ddev&amp;sort=ct">Calls</a></nobr></th><th class="vwbar"><nobr>Calls%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=isIgnoredUnserAgentRegex&amp;source=ddev&amp;sort=wt">Incl. Wall Time<br>(microsec)</a></nobr></th><th class="vwbar"><nobr>IWall%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=isIgnoredUnserAgentRegex&amp;source=ddev&amp;sort=cpu">Incl. CPU<br>(microsecs)</a></nobr></th><th class="vwbar"><nobr>ICpu%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=isIgnoredUnserAgentRegex&amp;source=ddev&amp;sort=mu">Incl.<br>MemUse<br>(bytes)</a></nobr></th><th class="vwbar"><nobr>IMemUse%</nobr></th><th class="vwbar"><nobr><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=isIgnoredUnserAgentRegex&amp;source=ddev&amp;sort=pmu">Incl.<br> PeakMemUse<br>(bytes)</a></nobr></th><th class="vwbar"><nobr>IPeakMemUse%</nobr></th></tr><tr bgcolor="#e0e0ff"><td><b><i><center>Current Function</center></i></b></td></tr><tr><td><a href="">isIgnoredUnserAgentRegex</a></td><td class="vbar">1</td>
<td class="vbar">3.2%</td>
<td class="vbbar">9,859</td>
<td class="vbbar">0.6%</td>
<td class="vbar">8,006</td>
<td class="vbar">0.5%</td>
<td class="vbar">3,393,248</td>
<td class="vbar">10.1%</td>
<td class="vbar">0</td>
<td class="vbar">0.0%</td>
</tr><tr bgcolor="#ffffff"><td style="text-align:right;color:blue">Exclusive Metrics  for Current Function</td><td class="vbar"></td><td class="vbar"></td><td type="Child" metric="wt" class="vbbar">24</td>
<td type="Child" metric="wt" class="vbbar">0.2%</td>
<td type="Child" metric="cpu" class="vbar">0</td>
<td type="Child" metric="cpu" class="vbar">0.0%</td>
<td type="Child" metric="mu" class="vbar">664</td>
<td type="Child" metric="mu" class="vbar">0.0%</td>
<td type="Child" metric="pmu" class="vbar">0</td>
<td type="Child" metric="pmu" class="vbar">N/A%</td>
</tr><tr bgcolor="#e0e0ff"><td><b><i><center>Parent function</center></i></b></td></tr><tr><td><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=main%28%29&amp;source=ddev">main()</a></td><td type="Parent" metric="ct" class="vbar">1</td>
<td type="Parent" metric="ct" class="vbar">100.0%</td>
<td type="Parent" metric="wt" class="vbbar">9,859</td>
<td type="Parent" metric="wt" class="vbbar">100.0%</td>
<td type="Parent" metric="cpu" class="vbar">8,006</td>
<td type="Parent" metric="cpu" class="vbar">100.0%</td>
<td type="Parent" metric="mu" class="vbar">3,393,248</td>
<td type="Parent" metric="mu" class="vbar">100.0%</td>
<td type="Parent" metric="pmu" class="vbar">0</td>
<td type="Parent" metric="pmu" class="vbar">N/A%</td>
</tr><tr bgcolor="#e0e0ff"><td><b><i><center>Child functions</center></i></b></td></tr><tr><td><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=implode&amp;source=ddev">implode</a></td><td type="Child" metric="ct" class="vbar">1</td>
<td type="Child" metric="ct" class="vbar">50.0%</td>
<td type="Child" metric="wt" class="vbbar">9,510</td>
<td type="Child" metric="wt" class="vbbar">96.5%</td>
<td type="Child" metric="cpu" class="vbar">8,006</td>
<td type="Child" metric="cpu" class="vbar">100.0%</td>
<td type="Child" metric="mu" class="vbar">3,392,048</td>
<td type="Child" metric="mu" class="vbar">100.0%</td>
<td type="Child" metric="pmu" class="vbar">0</td>
<td type="Child" metric="pmu" class="vbar">N/A%</td>
</tr><tr bgcolor="#e5e5e5"><td><a href="/xhprof/index.php?run=645a42885d6a7&amp;symbol=preg_match&amp;source=ddev">preg_match</a></td><td type="Child" metric="ct" class="vbar">1</td>
<td type="Child" metric="ct" class="vbar">50.0%</td>
<td type="Child" metric="wt" class="vbbar">325</td>
<td type="Child" metric="wt" class="vbbar">3.3%</td>
<td type="Child" metric="cpu" class="vbar">0</td>
<td type="Child" metric="cpu" class="vbar">0.0%</td>
<td type="Child" metric="mu" class="vbar">536</td>
<td type="Child" metric="mu" class="vbar">0.0%</td>
<td type="Child" metric="pmu" class="vbar">0</td>
<td type="Child" metric="pmu" class="vbar">N/A%</td>
</tr></tbody></table>